crates/oci/src/lib.rs

```diff
@@ -1,448 +1,39 @@
-
-        Ok(())
-    }
-
-    fn default_path() -> Result<PathBuf> {
-        Ok(dirs::config_dir()
-            .context("Cannot find configuration directory")?
-            .join("fermyon")
-            .join("registry-auth.json"))
-    }
-
-    /// Get the registry authentication for a given registry from the default location.
-    async fn get_auth_from_default(server: impl AsRef<str>) -> Result<RegistryAuth> {
-        let auths = Self::load_default().await?;
-        let encoded = match auths.auths.get(&server.as_ref().to_string()) {
-            Some(e) => e,
-            None => bail!(format!("no credentials stored for {}", server.as_ref())),
-        };
-
-        let bytes = base64::Engine::decode(&base64::engine::general_purpose::STANDARD, encoded)?;
-        let decoded = std::str::from_utf8(&bytes)?;
-        let parts: Vec<&str> = decoded.splitn(2, ':').collect();
-
-        tracing::trace!("Decoded registry credentials from the Spin configuration.");
-        Ok(RegistryAuth::Basic(
-            parts
-                .first()
-                .context("expected username as first element of the decoded auth")?
-                .to_string(),
-            parts
-                .get(1)
-                .context("expected secret as second element of the decoded auth")?
-                .to_string(),
-        ))
-    }
-
-    async fn load(p: &Path) -> Result<Self> {
-        let contents = tokio::fs::read_to_string(&p).await?;
-        serde_json::from_str(&contents)
-            .with_context(|| format!("cannot load authentication file {:?}", p))
-    }
-
-    async fn save(&self, p: &Path) -> Result<()> {
-        tokio::fs::write(&p, &serde_json::to_vec_pretty(&self)?)
-            .await
-            .with_context(|| format!("cannot save authentication file {:?}", p))
-    }
+mod auth;
+mod client;
+mod loader;
+
+pub use client::Client;
+
+/// Applies a heuristic to check if the given string "looks like" it could be
+/// an OCI reference.
+pub fn is_probably_oci_reference(maybe_oci: &str) -> bool {
+    // A relative file path such as foo/spin.toml will successfully
+    // parse as an OCI reference, because the parser infers the Docker
+    // registry and the `latest` version.  So if the registry resolves
+    // to Docker, but the source *doesn't* contain the string 'docker',
+    // we can guess this is likely a false positive.
+    //
+    // This could be fooled by, e.g., dockerdemo/spin.toml.  But we only
+    // go down this path if the file does not exist, and the chances of
+    // a user choosing a filename containing 'docker' THAT ALSO does not
+    // exist are A MILLION TO ONE...
+
+    // If it doesn't parse as a reference, it isn't a reference
+    let Ok(reference) = oci_distribution::Reference::try_from(maybe_oci) else {
+        return false;
+    };
+    // If it has an explicit tag, its probably a reference
+    if reference.tag() != Some("latest") || maybe_oci.ends_with(":latest") {
```

Comment:
@itowlson This whole function is mostly a verbose refactoring of your code, but it is tweaked slightly here to detect e.g. `actually/oci/reference:latest` as a reference.